### DISCUSSION TITLE
ESP32 Ethernet (Wemo emulation) doesn't form SSDP packet properly.

### DISCUSSION THREAD

**Discussion by sergshadrap** [arendst/Tasmota#16753] [Q&A]
Hi folks! Could you do me a favor  with my question?
Pocking around Wemo Emulation with ESP32 board  a few bugs where noticed, and one of them turns to be crucial for me.
All my home's Iot devices works under AutomationManager software and it can understand Tasmota devices. So my first intention was - running Tasmota on ESP32 Ethernet board in Wemo emulation mode with two relays on board. First of all I setup everything over wifi connection and all was fine-  AM Manager saw the Tasmota's devices.  At the second step I've switched the wifi off and connected the ESP board over Ethernet. The Tasmota's part was working well , but when I tried to look for Tasmota device form my AM Manager it couldn't find the device. Trying to figure out what is wrong with that pair I connected the Wireshark and full debug info in Tasmota device. When the AutomationManager starting to seek Iot devices it sends SSDP query and every devices response by advertising it's location . Tasmota's board response properly with it's location in wifi mode but not in Ethernet mode (it sends 0.0.0.0 as address).
I depicted the sample of the packet from Tasmota below:
Frame 225: 123 bytes on wire (984 bits), 123 bytes captured (984 bits) on interface \Device\NPF_{06FED1BC-87A0-4735-BB92-625CDE472606}, id 0
Ethernet II, Src: Espressi_35:ee:db (e8:31:cd:35:ee:db), Dst: Broadcast (ff:ff:ff:ff:ff:ff)
Internet Protocol Version 4, Src: 192.168.1.58, Dst: 0.0.0.0
User Datagram Protocol, Src Port: 1900, Dst Port: 0
Simple Service Discovery Protocol
    tasmota-35EED8-3800-eth ESP-WMO: WeMo Type 2, Response sent to 192.168.1.35:54625
        [Expert Info (Warning/Protocol): Illegal characters found in header name]
            [Illegal characters found in header name]
            [Severity level: Warning]
            [Group: Protocol]
Note: 192.168.1.35 - the AM Manager device , 192.168.1.58 - Tasmota's device
Could you check my assumption as well ?  Maybe I was wrong somewhere? Maybe it's a bug when the device is trying to read information from wifi connection only?

**Comment by s-hadinger** [comment]
It's a UDP broadcast packet, part of UPNP protocol. The destination address has no significance. If there is a difference in routing of UDP multicast, I suspect that your router is filtering broadcast/multicast somewhere.

I know that Alexa/Hue emulation works on Ethernet and it uses a similar UPNP discovery mechanism. I haven't tried though Wemo over Ethernet and can't try right now.

**Reply by sergshadrap** [reply]
Did you mean my router is filtering UDP multicast especially for Ethernet connections ? Could you suppose how I can check it?

**Reply by sergshadrap** [reply]
Any way I've checked all UDP traffic from the board , all broadcast packets goes for address 0.0.0.0 , that's barely right. They should be sending to 239.255.255.250. Wifi's packet goes in right way...

**Reply by barbudor** [reply]
Do you mean the packet is well formed (with the Multicast destination group) on Wifi but is missing the Multicast destination group on Ethernet ?
Strange as that would mean a problem with multicast at lower levels, not in Tasmota
A quick search on Google doesn't ring anything

**Reply by sergshadrap** [reply]
I risky to suppose the clue is somewhere in EthernetEvent , where wifi properties is hard coded to check only wifi info, so when wifi is on everything goes right way, but when wifi is off and ethernet is online only the system lost part of address info. Yep the packet is whole and completed, a router doesn't argue with them.

**Comment by s-hadinger** [comment]
Can you set log level to `weblog 3` and try again. Looking at the code, the destination address is supposed to be the source address of the initial udp packet

**Reply by sergshadrap** [reply]
Sorry for my dumbness, i don't pick up what did you mean under "Looking at the code" ? I didn't compile the master source yet.. 
Ok,I switched the extended reporting On in the log, even with weblog 4. 
I can see : 
11:04:59.997 -> 09:04:59.870 BRY: GC from 4115 to 3566 bytes, objects freed 2/40 (in 1 ms) - slots from 47/122 to 41/122
11:04:59.997 -> 09:04:59.875 RSL: STATE = {"Time":"2022-10-12T09:04:59","Uptime":"0T00:05:08","UptimeSec":308,"Heap":169,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":20,"MqttCount":0,"Berry":{"HeapUsed":3,"Objects":40},"POWER1":"ON","POWER2":"OFF"}
11:05:00.045 -> 09:04:59.906 RSL: SENSOR = {"Time":"2022-10-12T09:04:59","ESP32":{"Temperature":53.3},"TempUnit":"C"}
11:05:50.616 -> 09:05:50.482 UDP: Packet (119/520)
11:05:50.897 -> 09:05:50.783 UDP: Packet (119/506)
11:05:51.038 -> 09:05:50.934 UDP: Packet (119/504)
11:05:51.319 -> 09:05:51.184 UDP: Packet (119/440)
11:05:51.739 -> 09:05:51.634 UDP: Packet (119/449)
11:05:53.099 -> 09:05:52.984 UDP: Packet (119/492)
Could you navigate me more precisely I could answer your question? 

For more detailed information , there is extended log from start of the module:
10:59:51.904 -> rst:0x1 (POWERON_RESET),boot:0x12 (SPI_FAST_FLASH_BOOT)
10:59:51.904 -> configsip: 0, SPIWP:0xee
10:59:51.904 -> clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
10:59:51.951 -> mode:DOUT, clock div:2
10:59:51.951 -> load:0x3fff0030,len:184
10:59:51.951 -> load:0x40078000,len:12616
10:59:51.951 -> load:0x40080400,len:2916
10:59:51.951 -> entry 0x400805c4
10:59:51.999 -> 
10:59:52.047 -> 00:00:00.003 HDW: ESP32-D0WD-V3 (PSRAM)
10:59:52.047 -> 00:00:00.042 UFS: FlashFS mounted with 304 kB free
10:59:52.094 -> 00:00:00.090 CFG: Loaded from File, Count 132
10:59:52.142 -> 00:00:00.106 QPC: Count 1
10:59:52.142 -> 00:00:00.108 CFG: CR 385/699, Busy 0
10:59:52.142 -> 00:00:00.117 ROT: Mode 1
10:59:52.142 -> 00:00:00.131 BRY: GC from 3703 to 3071 bytes, objects freed 5/22 (in 0 ms) - slots from 34/61 to 23/61
10:59:52.236 -> 00:00:00.240 CFG: No '*.autoconf' file found
10:59:52.284 -> 00:00:00.246 BRY: GC from 4424 to 3735 bytes, objects freed 5/38 (in 1 ms) - slots from 54/76 to 35/76
10:59:52.284 -> 00:00:00.247 BRY: Berry initialized, RAM used=3735 bytes
10:59:52.284 -> 00:00:00.273 BRY: No 'preinit.be'
10:59:52.284 -> 00:00:00.277 SRC: Restart
10:59:52.284 -> 00:00:00.284 Project tasmota - Tasmota Version 12.1.1(tasmota)-2_0_4_1(2022-08-25T11:34:17)
10:59:53.718 -> 00:00:01.605 ETH: Attempting connection...
10:59:53.718 -> 00:00:01.607 ETH: Connected at 100Mbps Full Duplex
10:59:53.861 -> 00:00:01.749 BRY: No 'autoexec.be'
10:59:54.389 -> 00:00:02.240 ETH: Mac E8:31:CD:35:EE:DB, IPAddress 192.168.1.58, Hostname tasmota-35EED8-3800-eth
10:59:54.437 -> 00:00:02.298 HTP: Web server active on tasmota-35EED8-3800-eth with IP address 192.168.1.58
10:59:54.437 -> 00:00:02.301 UPP: Multicast (re)joined
10:59:55.054 -> 00:00:02.904 NTP: Sync time...
10:59:55.101 -> 00:00:02.951 RTC: UTC 2022-10-12T07:59:54, DST 2022-03-27T02:00:00, STD 2022-10-30T03:00:00
10:59:55.101 -> 08:59:54.001 RTC: Synced by NTP
10:59:55.385 -> 08:59:55.270 RSL: INFO1 = {"Info1":{"Module":"ESP32-DevKit","Version":"12.1.1(tasmota)","FallbackTopic":"cmnd/DVES_35EED8_fb/","GroupTopic":"cmnd/tasmotas/"}}
10:59:55.432 -> 08:59:55.285 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Ethernet":{"Hostname":"tasmota-35EED8-3800-eth","IPAddress":"192.168.1.58"}}}
10:59:55.432 -> 08:59:55.299 RSL: INFO3 = {"Info3":{"RestartReason":"Vbat power on reset","BootCount":54}}
10:59:55.432 -> 08:59:55.303 RSL: RESULT = {"POWER1":"ON"}
10:59:55.432 -> 08:59:55.305 RSL: POWER1 = ON
10:59:55.432 -> 08:59:55.307 RSL: RESULT = {"POWER2":"OFF"}
10:59:55.432 -> 08:59:55.309 RSL: POWER2 = OFF
10:59:55.432 -> 08:59:55.317 TFS: File 'autoexec.bat' not found
11:00:00.043 -> 08:59:59.893 QPC: Reset
11:00:00.043 -> 08:59:59.900 BRY: GC from 5326 to 3566 bytes, objects freed 25/40 (in 2 ms) - slots from 80/122 to 41/122
11:00:00.043 -> 08:59:59.904 RSL: STATE = {"Time":"2022-10-12T08:59:59","Uptime":"0T00:00:08","UptimeSec":8,"Heap":171,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":3,"Objects":40},"POWER1":"ON","POWER2":"OFF"}
11:00:00.043 -> 08:59:59.935 RSL: SENSOR = {"Time":"2022-10-12T08:59:59","ESP32":{"Temperature":53.3},"TempUnit":"C"}
11:00:01.998 -> 09:00:01.851 APP: Boot Count 54
11:00:03.130 -> 09:00:03.017 CFG: Saved, Count 133, Bytes 4096
11:00:05.036 -> 09:00:04.892 RSL: config = {"ip":"192.168.1.58","dn":"Tasmota","fn":["Tasmota","Tasmota2",null,null,null,null,null,null],"hn":"tasmota-35EED8-3800-eth","mac":"E831CD35EED8","md":"ESP32-DevKit","ty":0,"if":0,"ofln":"Offline","onln":"Online","state":["OFF","ON","TOGGLE","HOLD"],"sw":"12.1.1","t":"tasmota_35EED8","ft":"%prefix%/%topic%/","tp":["cmnd","stat","tele"],"rl":[1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"swc":[-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1],"swn":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],"btn":[0,0,0,0,0,0,0,0],"so":{"4":0,"11":0,"13":0,"17":0,"20":0,"30":0,"68":0,"73":0,"82":0,"114":0,"117":0},"lk":0,"lt_st":0,"sho":[0,0,0,0],"sht":[[0,0,0],[0,0,0],[0,0,0],[0,0,0]],"ver":1}
11:00:05.083 -> 09:00:04.973 RSL: sensors = {"sn":{"Time":"2022-10-12T09:00:04","ESP32":{"Temperature":53.3},"TempUnit":"C"},"ver":1}

If you meant be_udp_lib.h .... i still didn't figure it out..., but I see , the wifi is base for Ethernet things too , just why I thought about possible place for mistake.
